### DISCUSSION TITLE
Ability to highlight text by pattern

### DISCUSSION THREAD

**Discussion by failable** [zed-industries/zed#26200] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Can be helpful for reading codes. Don't necessary to be semantic.

### If applicable, add mockups / screenshots to help present your vision of the feature

<img width="1072" alt="image" src="https://github.com/zed-industries/community/assets/5475262/30b983bb-690a-490e-b201-7a78c8f179ac">

**Comment by JosephTLyons** [comment]
Hey @liebkne, can you explain this a bit more? Is the regex and color stored in some settings file?

**Comment by failable** [comment]
@JosephTLyons 

Hi, the regex or the pattern is interactively input by user, thought one should be also highlight symbol under cursor without inputing regex.

The colors might be built-in or configurable. When highlighting, colors might be selectable or automatically selected. 

I think It may be possible to just re-use the current search interface by adding a "Highlight matches", or use other interaction. 

Below is several demos:

- Highlight current symbol under cursor (Don't need to select colors, trigger by a shortcut simply)
   
  https://github.com/zed-industries/community/assets/5475262/e872016f-3aad-4dba-9723-79f19d21ed3e

- Highlight by regex and manually select a color (by calling command or shortcut)
   
   https://github.com/zed-industries/community/assets/5475262/faadd2b3-0e68-45c4-a0f5-0c84a5346cbd

- Select a region then highlight (by shortcut)

   https://github.com/zed-industries/community/assets/5475262/2d655c8c-0038-463f-a97a-eb1f2a7d8b94

Note that highlighting does not necessary mean selecting. But it will be helpful to have the ability turn highlights into selections like


https://github.com/zed-industries/community/assets/5475262/14716c8f-c483-4d06-800d-a64da7e5d0be

**Comment by JosephTLyons** [comment]
Ok, that all makes sense now. Appreciate the super thorough reply @liebkne!

**Comment by notpeter** [comment]
We recently shipped:
- https://github.com/zed-industries/zed/pull/24835

Which supports the simplest use case of this, since `selection_highlight` is by default set to true, when you select there will be a dim selection of all matches.

The emacs version seems more advanced, supporting regular expressions and other features.
Could you take a moment and enumerate (in words) the various ways you one can interact with the emacs feature.  Currently this feels really big, but I think it might be possible to scope something more reasonable.

**Reply by failable** [reply]
> Could you take a moment and enumerate (in words) the various ways you one can interact with the emacs feature.

Hello, does [this comment](https://github.com/zed-industries/zed/discussions/26200#discussioncomment-12410340) helps? 

There are several basic cases:

- Highlight current symbol under cursor (Don't need to select colors, trigger by a shortcut simply). This is similar to the PR you linked but even the cursor is moved the highlight is persistent until one calls the unhighlight command.
- Highlight by regex and manually select a color (by calling command or shortcut). 
- Select a region then highlight (by shortcut)
- Note that highlighting does not necessary mean selecting. But it will be helpful to have the ability turn highlights into selections.

For me, highlights are particular helpful when reading codes for tracking variables.
